File: nctreport.py
#!/usr/bin/env python 

"""
Maintain data for an clinical trials report
"""
__author__ = 'Rodney L. Summerscales'

import xmlutil

class ReportEntry:
    """ A text block for an entry in a trial registry """
    sentences = None
    name = None

    def __init__(self, node):
        """ create a list of sentences given an XML node containing a list of sentence
            elements """
        self.name = xmlutil.getNodeTagName(node)
        self.sentences = parseSentences(node)

    def getXML(self, doc):
        """ return an xml node with information for this entity """
        return createSentenceListNode(self.name, self.sentences, doc)

class Intervention:
    name = None
    description = None

    """ Intervention information from a trial registry """
    def __init__(self, node):
        self.name = []
        self.description = []
        nodes = node.getElementsByTagName('name')
        if len(nodes) > 0:
            self.name = parseSentences(nodes[0])
        nodes = node.getElementsByTagName('description')
        if len(nodes) > 0:
            self.description = parseSentences(nodes[0])

    def getXML(self, doc):
        node = doc.createElement('intervention')

        if len(self.name) > 0:
            node.appendChild(createSentenceListNode('name', self.name, doc))
        if len(self.description) > 0:
            node.appendChild(createSentenceListNode('description', self.description, doc))
        return node

class Outcome:
    """ Outcome information from a trial registry """
    primary = False
    name = None
    times = None
    description = None

    def __init__(self, node):
        self.name = []
        self.times = []
        self.description = []
        primaryStr = node.getAttribute('primary')
        if primaryStr != None and primaryStr.lower() == 'true':
            self.primary = True
        else:
            self.primary = False
        nodes = node.getElementsByTagName('name')
        if len(nodes) > 0:
            self.name = parseSentences(nodes[0])
        nodes = node.getElementsByTagName('description')
        if len(nodes) > 0:
            self.description = parseSentences(nodes[0])
        nodes = node.getElementsByTagName('times')
        if len(nodes) > 0:
            self.times = parseSentences(nodes[0])

    def getXML(self, doc):
        node = doc.createElement('outcome')
        if self.primary:
            node.setAttribute('primary', 'true')
        else:
            node.setAttribute('primary', 'false')

        if len(self.name) > 0:
            node.appendChild(createSentenceListNode('name', self.name, doc))
        if len(self.description) > 0:
            node.appendChild(createSentenceListNode('description', self.description, doc))
        if len(self.times) > 0:
            node.appendChild(createSentenceListNode('times', self.times, doc))
        return node



class Report:
    """ Data from a NCT report. """
    id = ''
    gender = ''
    minAge = ''
    maxAge = ''
    locations = None
    conditions = None
    eligibilityCriteria = None
    inclusionCriteria = None
    exclusionCriteria = None
    interventions = None
    outcomes = None

    def __init__(self, node):
        self.id = xmlutil.getTextFromNodeCalled('id', node)
        self.gender = xmlutil.getTextFromNodeCalled('gender', node)
        self.minAge = xmlutil.getTextFromNodeCalled('minAge', node)
        self.maxAge = xmlutil.getTextFromNodeCalled('maxAge', node)

        self.locations = []
        lcNodes = node.getElementsByTagName('location_countries')
        if len(lcNodes) > 0:
            cNodes = lcNodes[0].getElementsByTagName('country')
            for countryNode in cNodes:
                self.locations.append(xmlutil.getText(countryNode))

        self.conditions = []
        cNodes = node.getElementsByTagName('condition')
        for cNode in cNodes:
            self.conditions.append(ReportEntry(cNode))

        self.eligibilityCriteria = []
        ecNodes = node.getElementsByTagName('eligibility')
        if len(ecNodes) > 0:
            cNodes = ecNodes[0].getElementsByTagName('criteria')
            for cNode in cNodes:
                self.eligibilityCriteria.append(ReportEntry(cNode))

        self.inclusionCriteria = []
        icNodes = node.getElementsByTagName('inclusion')
        if len(icNodes) > 0:
            cNodes = icNodes[0].getElementsByTagName('criteria')
            for cNode in cNodes:
                self.inclusionCriteria.append(ReportEntry(cNode))

        self.exclusionCriteria = []
        ecNodes = node.getElementsByTagName('exclusion')
        if len(ecNodes) > 0:
            cNodes = ecNodes[0].getElementsByTagName('criteria')
            for cNode in cNodes:
                self.exclusionCriteria.append(ReportEntry(cNode))

        self.interventions = []
        iNodes = node.getElementsByTagName('intervention')
        for iNode in iNodes:
            self.interventions.append(Intervention(iNode))

        self.outcomes = []
        oNodes = node.getElementsByTagName('outcome')
        for oNode in oNodes:
            self.outcomes.append(Outcome(oNode))


    def getXML(self, doc):
        """ return an xml element containing information for a clinical report element"""
        node = doc.createElement('report')
        node.appendChild(xmlutil.createNodeWithTextChild(doc, 'id', self.id))
        node.appendChild(xmlutil.createNodeWithTextChild(doc, 'gender', self.gender))
        node.appendChild(xmlutil.createNodeWithTextChild(doc, 'minAge', self.minAge))
        node.appendChild(xmlutil.createNodeWithTextChild(doc, 'maxAge', self.maxAge))

        if len(self.locations) > 0:
            lcNode = doc.createElement('location_countries')
            node.appendChild(lcNode)
            for country in self.locations:
                lcNode.appendChild(xmlutil.createNodeWithTextChild(doc, 'country', country))

        for condition in self.conditions:
            node.appendChild(condition.getXML(doc))

        if len(self.eligibilityCriteria) > 0:
            ecNode = doc.createElement('eligibility')
            node.appendChild(ecNode)
            for ec in self.eligibilityCriteria:
                ecNode.appendChild(ec.getXML(doc))

        if len(self.inclusionCriteria) > 0:
            icNode = doc.createElement('inclusion')
            node.appendChild(icNode)
            for ic in self.inclusionCriteria:
                icNode.appendChild(ic.getXML(doc))

        if len(self.exclusionCriteria) > 0:
            ecNode = doc.createElement('exclusion')
            node.appendChild(ecNode)
            for ec in self.exclusionCriteria:
                ecNode.appendChild(ec.getXML(doc))

        for intervention in self.interventions:
            node.appendChild(intervention.getXML(doc))

        for outcome in self.outcomes:
            node.appendChild(outcome.getXML(doc))

        return node

File: src/ebm/NctReport.java
package ebm;

import java.util.*;
import org.jdom.Element;

public class NctReport {
	private class Criteria {
		Vector<Sentence> sentences = null;
		
		public Element getXMLElement(){
			Element cElement = new Element("criteria");
			for (Sentence sentence: sentences){
				cElement.addContent(sentence.getXMLElement());
			}
			
			return cElement;
		}

	}
	
	private class InterventionElement{
		Vector<Sentence> name = null;
		Vector<Sentence> description = null;
		
		public InterventionElement(Element iElement, LanguageModels models){
			Element nElement = iElement.getChild("intervention_name");
			if (nElement != null) {
				String text = nElement.getText();
				name = AnnotateText(text, models);
			}
			Element dElement = iElement.getChild("description");
			if (dElement != null) {
				String text = dElement.getText();
				description = AnnotateText(text, models);
			}

		}
		public Element getXMLElement(){
			Element iElement = new Element("intervention");
			
			if(name != null){
				Element e = new Element("name");
				for (Sentence sentence: name){
					e.addContent(sentence.getXMLElement());
				}
				iElement.addContent(e);
			}
			if(description != null){
				Element e = new Element("description");
				for (Sentence sentence: description){
					e.addContent(sentence.getXMLElement());
				}
				iElement.addContent(e);
			}

			return iElement;
		}
		
	}

	private class OutcomeElement{
		Vector<Sentence> name = null;
		Vector<Sentence> description = null;
		Vector<Sentence> times = null;
		boolean primaryOutcome = false;
		
		public OutcomeElement(Element element, LanguageModels models){
			primaryOutcome = element.getName().equalsIgnoreCase("primary_outcome");
			Element nElement = element.getChild("measure");
			if (nElement != null) {
				String text = nElement.getText();
				name = AnnotateText(text, models);
			}
			Element dElement = element.getChild("description");
			if (dElement != null) {
				String text = dElement.getText();
				description = AnnotateText(text, models);
			}
			Element tElement = element.getChild("time_frame");
			if (tElement != null) {
				String text = tElement.getText();
				times = AnnotateText(text, models);
			}
			

		}
		public Element getXMLElement(){
			Element oElement = new Element("outcome");
			if(primaryOutcome){
				oElement.setAttribute("primary", "true");
			} else {
				oElement.setAttribute("primary", "false");
			}
			
			if(name != null){
				Element e = new Element("name");
				for (Sentence sentence: name){
					e.addContent(sentence.getXMLElement());
				}
				oElement.addContent(e);
			}
			if(description != null){
				Element e = new Element("description");
				for (Sentence sentence: description){
					e.addContent(sentence.getXMLElement());
				}
				oElement.addContent(e);
			}
			if(times != null){
				Element e = new Element("times");
				for (Sentence sentence: times){
					e.addContent(sentence.getXMLElement());
				}
				oElement.addContent(e);
			}

			return oElement;
		}
		
	}

	Vector<InterventionElement> interventions = new Vector<InterventionElement>();
	Vector<OutcomeElement> outcomes = new Vector<OutcomeElement>();
	Vector<Criteria> inclusionCriteria = new Vector<Criteria>();
	Vector<Criteria> exclusionCriteria = new Vector<Criteria>();
	Vector<Criteria> allCriteria = new Vector<Criteria>();
	Vector<Sentence> conditions = new Vector<Sentence>();
	Vector<String> locations = new Vector<String>();
	String gender = "";
	String minAge = "";
	String maxAge = "";
	String id = "";
	
	/** Initialize given an xml tree containing an NCT report */
	@SuppressWarnings("unchecked")
	NctReport(Element root, LanguageModels models){
		try {
			Element idElement = root.getChild("id_info").getChild("nct_id");
			id = idElement.getText();
			
			// get conditions
			List<Element> cList = root.getChildren("condition");
			for(Element cElement: cList){
				conditions.addAll(AnnotateText(cElement.getText(), models));
			}
			
			// get locations
			Element lElement = root.getChild("location_countries");
			if(lElement != null){
				cList = lElement.getChildren("country");
				for(Element country: cList){
					locations.add(country.getText());
				}
			}
			
			// get eligibility criteria
			Element eElement = root.getChild("eligibility");
			if (eElement != null) {
				Element cElement = eElement.getChild("criteria");
				if (cElement != null) {
					Element tElement = cElement.getChild("textblock");
					if (tElement != null){
						String textblock = tElement.getText();
						if(textblock.contains("Exclusion Criteria:")){
							String[] blockParts = textblock.split("Exclusion Criteria:");
							String inclusionStr = blockParts[0].replace("Inclusion Criteria:", "");
							String exclusionStr = blockParts[1];
							inclusionCriteria = parseCriteria(inclusionStr, models);
							exclusionCriteria = parseCriteria(exclusionStr, models);
						} else {
							// just grab the entire block
							allCriteria = parseCriteria(textblock, models);
						}
					}
				}
				Element gElement = eElement.getChild("gender");
				if(gElement != null){
					gender = gElement.getText();
				}
				Element mElement = eElement.getChild("minimum_age");
				if(mElement != null){
					minAge = mElement.getText();
				}
				mElement = eElement.getChild("maximum_age");
				if(mElement != null){
					maxAge = mElement.getText();
				}
			}
			
			// get intervention information
			List<Element> iElementList = root.getChildren("intervention");
			for(Element iElement: iElementList){
				interventions.add(new InterventionElement(iElement, models));				
			}
			// get outcome information
			List<Element> oElementList = root.getChildren("primary_outcome");
			for(Element oElement: oElementList){
				outcomes.add(new OutcomeElement(oElement, models));				
			}
			oElementList = root.getChildren("secondary_outcome");
			for(Element oElement: oElementList){
				outcomes.add(new OutcomeElement(oElement, models));				
			}

		} catch (Exception e){
			System.out.println(e.getLocalizedMessage());
			e.printStackTrace();
			System.exit(-1);
		}
	}
	
	private Vector<Criteria> parseCriteria(String text, LanguageModels models){
		Vector<Criteria> cList = new Vector<Criteria>();
		text = text.replaceAll("               -", "vvvvv");
		String[] criteriaElements = text.split("-  ");
		for(String cText: criteriaElements){
			Criteria criteria = new Criteria();
			if(cText.contains("vvvvv")){
				// element contains sublist
				String[] elements = cText.split("vvvvv");
				if(elements.length > 1){
					cText = elements[0] + " " + elements[1];
					int i = 2;
					while(i < elements.length){
						cText += "; "+elements[i];
						i++;
					}
				}
				
			}
			criteria.sentences = AnnotateText(cText, models);
//			System.out.println(cText);
			if(criteria.sentences.size() > 0){
				cList.add(criteria);
			}
		}
		return cList;
	}
	
	private Vector<Sentence> AnnotateText(String text, LanguageModels models){
		Vector<RawTextFragment> fragmentList = new Vector<RawTextFragment>();
		String normalizedText = XmlUtil.normalizeText(text);
		fragmentList.add(new RawTextFragment(normalizedText));
		return models.AnnotateAndReturnSentences(fragmentList, "", "", true); 
	}
	
	/** return an XML element containing the abstract */
	public Element getXMLElement(){
		Element aElement = new Element("report");
		Element idElement = new Element("id");
		idElement.addContent(id);
		aElement.addContent(idElement);
		
		if(gender.length() > 0){
			Element gElement = new Element("gender");
			aElement.addContent(gElement);
			gElement.addContent(gender);
		}
		if(minAge.length() > 0){
			Element mElement = new Element("minAge");
			aElement.addContent(mElement);
			mElement.addContent(minAge);
		}
		if(maxAge.length() > 0){
			Element mElement = new Element("maxAge");
			aElement.addContent(mElement);
			mElement.addContent(maxAge);
		}
		
		if(locations.size() > 0) {
			Element lElement = new Element("location_countries");
			aElement.addContent(lElement);
			for(String country: locations){
				Element cElement = new Element("country");
				cElement.addContent(country);
				lElement.addContent(cElement);
			}
		}
		
		for(Sentence sentence: conditions){
			Element cElement = new Element("condition");
			cElement.addContent(sentence.getXMLElement());
			aElement.addContent(cElement);
		}
		
		if(inclusionCriteria.size() > 0){
			Element iElement = new Element("inclusion");
			aElement.addContent(iElement);
			for (Criteria criteria: inclusionCriteria){
				iElement.addContent(criteria.getXMLElement());
			}
		}
		if(exclusionCriteria.size() > 0){
			Element eElement = new Element("exclusion");
			aElement.addContent(eElement);
			for (Criteria criteria: exclusionCriteria){
				eElement.addContent(criteria.getXMLElement());
			}
		}
		if(allCriteria.size() > 0){
			Element eElement = new Element("eligibility");
			aElement.addContent(eElement);
			for (Criteria criteria: allCriteria){
				eElement.addContent(criteria.getXMLElement());
			}
		}

		for(InterventionElement iElement: interventions){
			aElement.addContent(iElement.getXMLElement());
		}
		for(OutcomeElement oElement: outcomes){
			aElement.addContent(oElement.getXMLElement());
		}
	
		return aElement;
	}

}